File: backend/app/routers/categories.py
from fastapi import APIRouter, HTTPException, status
from typing import List, Dict, Any
from datetime import datetime
from pydantic import BaseModel
import json
import os

router = APIRouter()

# File path to store category configuration
DATA_DIR = os.path.join(os.path.dirname(__file__), '..', '..', 'data')
CATEGORY_CONFIG_FILE = os.path.join(DATA_DIR, 'category_config.json')
ACTIVITY_DATA_DIR = os.path.join(os.path.dirname(__file__), '..', '..', '..', 'data-collector', 'activity_data')

# Pydantic models
class Application(BaseModel):
    name: str

class Category(BaseModel):
    id: str
    name: str
    icon: str
    color: str
    applications: List[str]

class CategoriesConfig(BaseModel):
    categories: List[Category]

# Default categories
DEFAULT_CATEGORIES = [
    {
        'id': 'productivity',
        'name': 'Productivity',
        'icon': 'Work',
        'color': '#4caf50',
        'applications': [
            'Visual Studio Code',
            'IntelliJ IDEA',
            'Eclipse',
            'PyCharm',
            'Sublime Text',
            'Notepad++',
            'Microsoft Excel',
            'Microsoft Word',
            'Microsoft PowerPoint',
            'Google Chrome',
            'Firefox',
            'Edge',
            'Notepad'
        ]
    },
    {
        'id': 'communication',
        'name': 'Communication',
        'icon': 'People',
        'color': '#2196f3',
        'applications': [
            'Microsoft Teams',
            'Slack',
            'Zoom',
            'Discord',
            'Skype',
            'Outlook',
            'Gmail',
            'Thunderbird'
        ]
    },
    {
        'id': 'break',
        'name': 'Break',
        'icon': 'Coffee',
        'color': '#ff9800',
        'applications': [
            'YouTube',
            'Spotify',
            'Netflix',
            'Steam',
            'Epic Games',
            'Twitter',
            'Facebook',
            'Instagram',
            'Reddit'
        ]
    }
]

def ensure_data_directory():
    """Ensure the data directory exists"""
    os.makedirs(DATA_DIR, exist_ok=True)

def load_categories() -> Dict[str, Any]:
    """Load categories from file or return defaults"""
    try:
        if os.path.exists(CATEGORY_CONFIG_FILE):
            with open(CATEGORY_CONFIG_FILE, 'r', encoding='utf-8') as f:
                return json.load(f)
        else:
            # Create default config file
            ensure_data_directory()
            default_config = {'categories': DEFAULT_CATEGORIES}
            save_categories(default_config)
            return default_config
    except Exception as e:
        print(f"Error loading categories: {e}")
        return {'categories': DEFAULT_CATEGORIES}

def save_categories(data: Dict[str, Any]) -> bool:
    """Save categories to file"""
    try:
        ensure_data_directory()
        with open(CATEGORY_CONFIG_FILE, 'w', encoding='utf-8') as f:
            json.dump(data, f, indent=2)
        return True
    except Exception as e:
        print(f"Error saving categories: {e}")
        return False

def get_category_for_app(app_name: str) -> str:
    """Get the category for a specific application"""
    categories = load_categories()
    for category in categories.get('categories', []):
        if app_name in category.get('applications', []):
            return category['id']
    return 'productivity'  # Default category

@router.get("/categories")
async def get_categories():
    """Get all categories"""
    try:
        data = load_categories()
        return data
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error loading categories: {str(e)}"
        )

@router.post("/categories")
async def update_categories(config: CategoriesConfig):
    """Update categories configuration"""
    try:
        # Convert to dict and add timestamp
        data = config.dict()
        data['updated_at'] = datetime.now().isoformat()
        
        if save_categories(data):
            return {
                'message': 'Categories updated successfully',
                'data': data
            }
        else:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to save categories"
            )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error updating categories: {str(e)}"
        )

@router.get("/categories/{category_id}")
async def get_category(category_id: str):
    """Get a specific category"""
    try:
        data = load_categories()
        for category in data.get('categories', []):
            if category['id'] == category_id:
                return category
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Category not found"
        )
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error getting category: {str(e)}"
        )

@router.get("/categories/app/{app_name}")
async def get_app_category(app_name: str):
    """Get the category for a specific application"""
    try:
        category_id = get_category_for_app(app_name)
        return {
            'app_name': app_name,
            'category': category_id
        }
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error getting app category: {str(e)}"
        )

@router.get("/applications")
async def get_all_applications():
    """Get all unique applications from activity data"""
    try:
        applications = set()
        
        # Get applications from activity data files
        if os.path.exists(ACTIVITY_DATA_DIR):
            for filename in os.listdir(ACTIVITY_DATA_DIR):
                if filename.endswith('.jsonl'):
                    filepath = os.path.join(ACTIVITY_DATA_DIR, filename)
                    try:
                        with open(filepath, 'r', encoding='utf-8') as f:
                            for line in f:
                                try:
                                    entry = json.loads(line)
                                    # Get apps from main apps array
                                    if 'apps' in entry:
                                        for app in entry['apps']:
                                            if 'title' in app and app['title']:
                                                applications.add(app['title'])
                                    # Get apps from background apps
                                    if 'backgroundApps' in entry and 'apps' in entry['backgroundApps']:
                                        for app in entry['backgroundApps']['apps']:
                                            if 'title' in app and app['title']:
                                                applications.add(app['title'])
                                except json.JSONDecodeError:
                                    continue
                    except Exception as e:
                        print(f"Error reading file {filename}: {e}")
                        continue
        
        # Also add applications from categories (to ensure configured apps are shown)
        categories_data = load_categories()
        for category in categories_data.get('categories', []):
            applications.update(category.get('applications', []))
        
        # Filter out empty strings and system apps
        filtered_apps = [app for app in applications if app and app.strip() and not app.startswith('System')]
        
        return {
            'applications': sorted(list(filtered_apps))
        }
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error getting applications: {str(e)}"
        )

@router.post("/categories/reset")
async def reset_categories():
    """Reset categories to default"""
    try:
        data = {
            'categories': DEFAULT_CATEGORIES,
            'updated_at': datetime.now().isoformat()
        }
        if save_categories(data):
            return {
                'message': 'Categories reset to default',
                'data': data
            }
        else:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to reset categories"
            )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error resetting categories: {str(e)}"
        )

File: backend-express/routes/categories.js
const express = require('express');
const router = express.Router();
const fs = require('fs').promises;
const path = require('path');

// Path to category config file
const CATEGORY_CONFIG_FILE = path.join(__dirname, '../data/category_config.json');
const ACTIVITY_DATA_DIR = path.join(__dirname, '../../data-collector/activity_data');

// Default categories
const DEFAULT_CATEGORIES = [
  {
    id: 'productivity',
    name: 'Productivity',
    icon: 'Work',
    color: '#4caf50',
    applications: [
      'Visual Studio Code',
      'IntelliJ IDEA',
      'Eclipse',
      'PyCharm',
      'Sublime Text',
      'Notepad++',
      'Microsoft Excel',
      'Microsoft Word',
      'Microsoft PowerPoint',
      'Google Chrome',
      'Firefox',
      'Edge',
      'Notepad'
    ]
  },
  {
    id: 'communication',
    name: 'Communication',
    icon: 'People',
    color: '#2196f3',
    applications: [
      'Microsoft Teams',
      'Slack',
      'Discord',
      'Zoom',
      'Skype',
      'Microsoft Outlook',
      'Gmail',
      'Thunderbird'
    ]
  },
  {
    id: 'break',
    name: 'Break',
    icon: 'Coffee',
    color: '#ff9800',
    applications: [
      'YouTube',
      'Netflix',
      'Spotify',
      'VLC Media Player',
      'Windows Media Player',
      'Steam',
      'Epic Games Launcher'
    ]
  }
];

// Helper function to ensure data directory exists
async function ensureDataDir() {
  const dataDir = path.dirname(CATEGORY_CONFIG_FILE);
  try {
    await fs.access(dataDir);
  } catch (error) {
    await fs.mkdir(dataDir, { recursive: true });
  }
}

// Helper function to read categories
async function readCategories() {
  try {
    const data = await fs.readFile(CATEGORY_CONFIG_FILE, 'utf8');
    const config = JSON.parse(data);
    return config.categories || DEFAULT_CATEGORIES;
  } catch (error) {
    if (error.code === 'ENOENT') {
      // File doesn't exist, create with default categories
      await ensureDataDir();
      await writeCategories(DEFAULT_CATEGORIES);
      return DEFAULT_CATEGORIES;
    }
    throw error;
  }
}

// Helper function to write categories
async function writeCategories(categories) {
  await ensureDataDir();
  const config = { 
    categories,
    last_updated: new Date().toISOString()
  };
  await fs.writeFile(CATEGORY_CONFIG_FILE, JSON.stringify(config, null, 2), 'utf8');
}

// Helper function to get all applications from activity data
async function getAllApplications() {
  const applications = new Set();
  
  try {
    const files = await fs.readdir(ACTIVITY_DATA_DIR);
    const jsonlFiles = files.filter(f => f.endsWith('.jsonl'));
    
    for (const file of jsonlFiles) {
      const filePath = path.join(ACTIVITY_DATA_DIR, file);
      const content = await fs.readFile(filePath, 'utf8');
      const lines = content.trim().split('\n').filter(line => line.trim());
      
      for (const line of lines) {
        try {
          const entry = JSON.parse(line);
          
          // Get from main apps
          if (entry.apps && Array.isArray(entry.apps)) {
            for (const app of entry.apps) {
              if (app.title && app.title.trim()) {
                applications.add(app.title.trim());
              }
            }
          }
          
          // Get from background apps
          if (entry.backgroundApps && entry.backgroundApps.apps && Array.isArray(entry.backgroundApps.apps)) {
            for (const app of entry.backgroundApps.apps) {
              if (app.title && app.title.trim()) {
                applications.add(app.title.trim());
              }
            }
          }
        } catch (parseError) {
          // Skip invalid JSON lines
          continue;
        }
      }
    }
  } catch (error) {
    console.error('Error reading activity data:', error);
  }
  
  // Filter out system apps and empty strings
  const systemApps = ['Windows Shell Experience Host', 'Windows Explorer', 'System'];
  const filteredApps = Array.from(applications).filter(app => 
    app && !systemApps.includes(app)
  );
  
  return filteredApps.sort();
}

// GET /api/categories - Get all categories
router.get('/categories', async (req, res) => {
  try {
    const categories = await readCategories();
    res.json({ categories });
  } catch (error) {
    console.error('Error reading categories:', error);
    res.status(500).json({ error: 'Failed to read categories' });
  }
});

// GET /api/applications - Get all available applications
router.get('/applications', async (req, res) => {
  try {
    const applications = await getAllApplications();
    res.json({ applications });
  } catch (error) {
    console.error('Error getting applications:', error);
    res.status(500).json({ error: 'Failed to get applications' });
  }
});

// POST /api/categories - Update all categories
router.post('/categories', async (req, res) => {
  try {
    const { categories } = req.body;
    
    if (!categories || !Array.isArray(categories)) {
      return res.status(400).json({ error: 'Invalid categories data' });
    }

    await writeCategories(categories);
    res.json({ 
      message: 'Categories updated successfully',
      categories 
    });
  } catch (error) {
    console.error('Error updating categories:', error);
    res.status(500).json({ error: 'Failed to update categories' });
  }
});

// POST /api/categories/reset - Reset to default categories
router.post('/categories/reset', async (req, res) => {
  try {
    await writeCategories(DEFAULT_CATEGORIES);
    res.json({ 
      message: 'Categories reset to defaults',
      categories: DEFAULT_CATEGORIES 
    });
  } catch (error) {
    console.error('Error resetting categories:', error);
    res.status(500).json({ error: 'Failed to reset categories' });
  }
});

module.exports = router;
